### PR TITLE
improve: series.Filter + drift.openPosition + modify openPosition behavior + fix consts

### PR DESCRIPTION
**Pull request by zenixls2**
feature: add series.filter
 fix: fixedpoint.Four
 fix: modifiable for embedded fields
 refactor: change drift to use openPosition
 refactor: modify openPosition behavior

### CODE REVIEW

**Comment by zenixls2** [comment]
TODO: for config printing, embedded fields should also be printed out.

**Comment by zenixls2** [comment]
TODO: FieldByIndexErr for 1.17

File: pkg/bbgo/order_executor_general.go [hunk 1]
```diff
@@ -246,24 +245,34 @@ func (e *GeneralOrderExecutor) OpenPosition(ctx context.Context, options OpenPos
...
-		submitOrder.Type = types.OrderTypeMarket
-	} else if options.LimitOrder {
+	if options.LimitOrder {
 		submitOrder.Type = types.OrderTypeLimit
 		submitOrder.Price = price
 	}
 
 	quantity := options.Quantity
 
+	market, ok := e.session.Market(e.symbol)
```
Comment (c9s) [inline pkg/bbgo/order_executor_general.go:255]: the market is already defined e.position.Market

File: pkg/bbgo/order_executor_general.go [hunk 2]
```diff
@@ -246,24 +245,34 @@ func (e *GeneralOrderExecutor) OpenPosition(ctx context.Context, options OpenPos
...
+	if !ok {
+		return nil, errors.New("cannot find market with symbol " + e.symbol)
+	}
+
 	if options.Long {
 		if quantity.IsZero() {
 			quoteQuantity, err := CalculateQuoteQuantity(ctx, e.session, e.position.QuoteCurrency, options.Leverage)
+			if quoteQuantity.IsZero() {
+				log.Warnf("dust quantity: %v", quantity)
+				return nil, nil
```
Comment (c9s) [inline pkg/bbgo/order_executor_general.go:265]: I think this should be error

File: pkg/bbgo/order_executor_general.go [hunk 3]
```diff
@@ -222,7 +219,9 @@ type OpenPositionOptions struct {
 	Tags  []string         `json:"-" yaml:"-"`
 }
 
-func (e *GeneralOrderExecutor) OpenPosition(ctx context.Context, options OpenPositionOptions) error {
+var Delta fixedpoint.Value = fixedpoint.NewFromFloat(0.005)
```
Comment (c9s) [inline pkg/bbgo/order_executor_general.go:222]: rename to quantityReduceDelta? and add a doc comment for this

File: pkg/bbgo/order_executor_general.go [hunk 4]
```diff
@@ -274,22 +283,35 @@ func (e *GeneralOrderExecutor) OpenPosition(ctx context.Context, options OpenPos
 		submitOrder.Side = types.SideTypeBuy
 		submitOrder.Quantity = quantity
 
-		Notify("Opening %s long position with quantity %f at price %f", e.position.Symbol, quantity.Float64(), price.Float64())
-		createdOrder, err2 := e.SubmitOrders(ctx, submitOrder)
-		if err2 != nil {
-			return err2
+		Notify("Opening %s long position with quantity %v at price %v", e.position.Symbol, quantity, price)
```
Comment (c9s) [inline pkg/bbgo/order_executor_general.go:286]: Call this for the first time:

createdOrder, err := e.SubmitOrders(ctx, submitOrder)

 if err != nil then we should call reduceQuantityAndSubmitOrders() method

File: pkg/bbgo/order_executor_general.go [hunk 5]
```diff
@@ -274,22 +283,35 @@ func (e *GeneralOrderExecutor) OpenPosition(ctx context.Context, options OpenPos
...
-
-		log.Infof("created order: %+v", createdOrder)
-		return nil
 	} else if options.Short {
 		if quantity.IsZero() {
 			var err error
 			quantity, err = CalculateBaseQuantity(e.session, e.position.Market, price, quantity, options.Leverage)
+			if quantity.IsZero() {
+				log.Warnf("dust quantity: %v", quantity)
+				return nil, nil
```
Comment (c9s) [inline pkg/bbgo/order_executor_general.go:305]: dust quantity should be error

File: pkg/bbgo/order_executor_general.go [hunk 6]
```diff
@@ -300,17 +322,22 @@ func (e *GeneralOrderExecutor) OpenPosition(ctx context.Context, options OpenPos
 		submitOrder.Side = types.SideTypeSell
 		submitOrder.Quantity = quantity
 
-		Notify("Opening %s short position with quantity %f at price %f", e.position.Symbol, quantity.Float64(), price.Float64())
-		createdOrder, err2 := e.SubmitOrders(ctx, submitOrder)
-		if err2 != nil {
-			return err2
+		Notify("Opening %s short position with quantity %v at price %v", e.position.Symbol, quantity, price)
+		for {
```
Comment (c9s) [inline pkg/bbgo/order_executor_general.go:326]: reduceQuantityAndSubmitOrders can be reused here

File: pkg/dynamic/field.go [hunk 7]
```diff
@@ -29,11 +30,20 @@ func LookupSymbolField(rs reflect.Value) (string, bool) {
 
 // Used by bbgo/interact_modify.go
 func GetModifiableFields(val reflect.Value, callback func(tagName, name string)) {
+	if val.Kind() == reflect.Ptr {
+		val = val.Elem()
+	}
+	if !val.IsValid() {
+		return
+	}
 	for i := 0; i < val.Type().NumField(); i++ {
```
Comment (c9s) [inline pkg/dynamic/field.go:39]: cache this val.Type() in valueType variable?

File: pkg/dynamic/field_test.go [hunk 8]
```diff
@@ -24,7 +35,7 @@ func TestGetModifiableFields(t *testing.T) {
 		assert.NotEqual(t, name, "Field2")
 		assert.NotEqual(t, tagName, "field3")
 		assert.NotEqual(t, name, "Field3")
-
+		fmt.Println(tagName, name)
```
Comment (c9s) [inline pkg/dynamic/field_test.go:38]: in test case, use t.Logf

File: pkg/types/indicator.go [hunk 9]
```diff
@@ -997,6 +998,51 @@ func Rolling(a Series, window int) *RollingResult {
 	return &RollingResult{a, window}
 }
 
+type FilterResult struct {
```
Comment (c9s) [inline pkg/types/indicator.go:1001]: move this to filter.go ?

File: pkg/types/indicator.go [hunk 10]
```diff
@@ -112,6 +112,7 @@ type SeriesExtend interface {
 	Softmax(window int) SeriesExtend
 	Entropy(window int) float64
 	CrossEntropy(b Series, window int) float64
+	Filter(b func(i int, value float64) bool, length int) SeriesExtend
```
Comment (c9s) [inline pkg/types/indicator.go:115]: can to add a doc comment for this Filter method?

Comment (c9s) [reply pkg/types/indicator.go:115]: for example, how does it do filtering?

File: pkg/strategy/drift/strategy.go [hunk 11]
```diff
@@ -377,9 +383,9 @@ func (s *Strategy) initTickerFunctions(ctx context.Context) {
 			exitLongCondition := s.buyPrice > 0 && (s.buyPrice*(1.-stoploss) >= pricef ||
 				s.trailingCheck(pricef, "long"))
 			if exitShortCondition || exitLongCondition {
-				log.Infof("Close position by orderbook changes")
-				s.positionLock.Unlock()
-				_ = s.ClosePosition(ctx, fixedpoint.One)
+				if s.ClosePosition(ctx, fixedpoint.One) {
+					log.Infof("Close position by orderbook changes")
```
Comment (c9s) [inline pkg/strategy/drift/strategy.go:387]: console logs always starts with lower case

File: pkg/bbgo/order_executor_general.go (hunk 1 above)
Comment (zenixls2) [reply pkg/bbgo/order_executor_general.go]: done

File: pkg/bbgo/order_executor_general.go (hunk 2 above)
Comment (zenixls2) [reply pkg/bbgo/order_executor_general.go]: done

File: pkg/bbgo/order_executor_general.go (hunk 3 above)
Comment (zenixls2) [reply pkg/bbgo/order_executor_general.go]: done

File: pkg/bbgo/order_executor_general.go (hunk 4 above)
Comment (zenixls2) [reply pkg/bbgo/order_executor_general.go:294]: done

File: pkg/bbgo/order_executor_general.go (hunk 5 above)
Comment (zenixls2) [reply pkg/bbgo/order_executor_general.go]: done

File: pkg/bbgo/order_executor_general.go (hunk 6 above)
Comment (zenixls2) [reply pkg/bbgo/order_executor_general.go]: done

File: pkg/dynamic/field.go (hunk 7 above)
Comment (zenixls2) [reply pkg/dynamic/field.go]: done

File: pkg/dynamic/field_test.go (hunk 8 above)
Comment (zenixls2) [reply pkg/dynamic/field_test.go]: removed

File: pkg/types/indicator.go (hunk 9 above)
Comment (zenixls2) [reply pkg/types/indicator.go]: done

File: pkg/types/indicator.go (hunk 10 above)
Comment (zenixls2) [reply pkg/types/indicator.go:115]: added in filter.go

File: pkg/strategy/drift/strategy.go (hunk 11 above)
Comment (zenixls2) [reply pkg/strategy/drift/strategy.go]: done

**Review by c9s — APPROVED** [review]

**Pull request merged**